File: clients/windows/Mailcal.Tests/ManualServerFieldTests.cs
// The manual form's port and connection security, as arithmetic: the WinUI-free half of the rule
// in docs/account-autodetect.md. The Windows twin of the same suite on the other three clients.
// What is load-bearing here is that the picker fills the port until the user takes it over, and
// that a port they typed is never overwritten: a server on a non-standard port is the reason the
// manual form exists at all.

using Allodia.Mailcal.Services;
using uniffi.mailcal_bindings;
using Xunit;

namespace Allodia.Mailcal.Tests;

public class ManualServerFieldTests
{
    // The standard ports, as the core answers them. They are written out here because this
    // assembly loads no cdylib; that they are the SAME numbers the core dials is pinned in
    // `mailcal-bindings`'s own suite, which is the only place that can assert it.
    private static ManualServerField Imap() =>
        new(security => security == ConnectionSecurity.ImplicitTls ? (ushort)993 : (ushort)143);

    private static ManualServerField Smtp() =>
        new(security => security == ConnectionSecurity.ImplicitTls ? (ushort)465 : (ushort)587);

    [Fact]
    public void AFreshFieldOffersTheStandardSecurePort()
    {
        Assert.Equal("993", Imap().Port);
        Assert.Equal("465", Smtp().Port);
        Assert.Equal(ConnectionSecurity.ImplicitTls, Imap().Security);
    }

    [Fact]
    public void ThePortFollowsTheSecurityPickerWhileItIsStillOurs()
    {
        var imap = Imap();
        imap.ChooseSecurity(ConnectionSecurity.StartTls);
        Assert.Equal("143", imap.Port);
        imap.ChooseSecurity(ConnectionSecurity.ImplicitTls);
        Assert.Equal("993", imap.Port);

        var smtp = Smtp();
        smtp.ChooseSecurity(ConnectionSecurity.StartTls);
        Assert.Equal("587", smtp.Port);
    }

    [Fact]
    public void APortTypedByHandIsNeverOverwrittenByThePicker()
    {
        var field = Imap();
        field.TypePort("1143");

        field.ChooseSecurity(ConnectionSecurity.StartTls);
        Assert.Equal("1143", field.Port);
        field.ChooseSecurity(ConnectionSecurity.ImplicitTls);
        Assert.Equal("1143", field.Port);
        Assert.False(field.FollowsSecurity);
    }

    [Fact]
    public void ClearingThePortHandsItBackToThePicker()
    {
        var field = Imap();
        field.TypePort("1143");
        field.TypePort("   ");

        Assert.True(field.FollowsSecurity);
        Assert.Equal("993", field.Port);
        field.ChooseSecurity(ConnectionSecurity.StartTls);
        Assert.Equal("143", field.Port);
    }

    [Fact]
    public void TheDialAddressCarriesTheHostAndThePortTogether()
    {
        var field = Imap();
        field.ChooseSecurity(ConnectionSecurity.StartTls);
        Assert.Equal("imap.example.net:143", field.Dial("imap.example.net"));

        field.TypePort("1143");
        Assert.Equal("127.0.0.1:1143", field.Dial("127.0.0.1"));
        Assert.Equal("127.0.0.1:1143", field.Dial("  127.0.0.1  "));
    }

    [Fact]
    public void APortAlreadyTypedIntoTheHostFieldWins()
    {
        // Two ports on one server would be a contradiction, and the one beside the name is the
        // one the user can see.
        var field = Imap();
        field.TypePort("1143");
        Assert.Equal("127.0.0.1:1025", field.Dial("127.0.0.1:1025"));
    }

    [Fact]
    public void AnEmptyHostStaysEmptySoTheConnectGateStillRefusesIt()
    {
        Assert.Equal(string.Empty, Imap().Dial("   "));
    }

    [Fact]
    public void ADetectedRouteBringsItsOwnPortAndStopsFollowingThePicker()
    {
        var field = Imap();
        field.AdoptDetected("imap.example.net:1993", ConnectionSecurity.StartTls);

        Assert.Equal("1993", field.Port);
        Assert.False(field.FollowsSecurity);
        Assert.Equal(ConnectionSecurity.StartTls, field.Security);
    }

    [Fact]
    public void ADetectedRouteWithoutAPortShowsTheStandardOneForWhatWasDetected()
    {
        var field = Imap();
        field.AdoptDetected("imap.example.net", ConnectionSecurity.StartTls);

        Assert.Equal("143", field.Port);
        Assert.True(field.FollowsSecurity);
    }

    // Mirrors the core's own split (mailcal-account's `host_and_addr`), deliberately: the core
    // decides the address actually dialled, so a client that split differently would show one
    // port and connect to another. A bare IPv6 literal is not a server either side accepts.
    [Theory]
    [InlineData("imap.example.net", "imap.example.net", "")]
    [InlineData("imap.example.net:993", "imap.example.net", "993")]
    [InlineData("host:", "host:", "")]
    [InlineData("host:abc", "host:abc", "")]
    [InlineData("::1", ":", "1")]
    public void OnlyAnAllDigitTailIsAPort(string input, string host, string port)
    {
        var split = ManualServerField.SplitHost(input);
        Assert.Equal(host, split.Host);
        Assert.Equal(port, split.Port);
    }
}

File: clients/apple/Packages/MailcalKit/Tests/MailcalKitTests/ManualServerFieldTests.swift
// The manual form's port and connection-security rule, as arithmetic. The Apple client's copy of
// the suite every client carries for `ManualServerField` (docs/account-autodetect.md).
//
// What is load-bearing: the picker fills the port until the user takes it over, and a port they
// typed is never overwritten. A server on a non-standard port is the reason the manual form
// exists, so a picker that resets it would defeat the form.

import MailcalBindings
import Testing

@testable import MailcalUI

@Suite struct ManualServerFieldTests {
    @Test func aFreshFieldOffersTheStandardSecurePort() {
        #expect(ManualServerField(.imap).port == "993")
        #expect(ManualServerField(.smtp).port == "465")
        #expect(ManualServerField(.imap).security == .implicitTls)
    }

    @Test func thePortFollowsThePickerWhileItIsStillOurs() {
        var imap = ManualServerField(.imap)
        imap.choose(.startTls)
        #expect(imap.port == "143")
        imap.choose(.implicitTls)
        #expect(imap.port == "993")

        var smtp = ManualServerField(.smtp)
        smtp.choose(.startTls)
        #expect(smtp.port == "587")
    }

    @Test func aPortTypedByHandIsNeverOverwrittenByThePicker() {
        var field = ManualServerField(.imap)
        field.typePort("1143")

        field.choose(.startTls)
        #expect(field.port == "1143")
        field.choose(.implicitTls)
        #expect(field.port == "1143")
        #expect(!field.followsSecurity)
    }

    @Test func clearingThePortHandsItBackToThePicker() {
        var field = ManualServerField(.imap)
        field.typePort("1143")
        field.typePort("   ")

        #expect(field.followsSecurity)
        #expect(field.port == "993")
        field.choose(.startTls)
        #expect(field.port == "143")
    }

    @Test func theDialAddressCarriesTheHostAndThePortTogether() {
        var field = ManualServerField(.imap)
        field.choose(.startTls)
        #expect(field.dial("imap.example.net") == "imap.example.net:143")

        field.typePort("1143")
        #expect(field.dial("127.0.0.1") == "127.0.0.1:1143")
        #expect(field.dial("  127.0.0.1  ") == "127.0.0.1:1143")
    }

    @Test func aPortAlreadyTypedIntoTheHostFieldWins() {
        var field = ManualServerField(.imap)
        field.typePort("1143")
        #expect(field.dial("127.0.0.1:1025") == "127.0.0.1:1025")
    }

    @Test func anEmptyHostStaysEmptySoTheConnectGateStillRefusesIt() {
        #expect(ManualServerField(.imap).dial("   ").isEmpty)
    }

    @Test func aDetectedRouteBringsItsOwnPortAndStopsFollowingThePicker() {
        var field = ManualServerField(.imap)
        field.adoptDetected(host: "imap.example.net:1993", security: .startTls)

        #expect(field.port == "1993")
        #expect(!field.followsSecurity)
        #expect(field.security == .startTls)
    }

    @Test func aDetectedRouteWithoutAPortShowsTheStandardOneForWhatWasDetected() {
        var field = ManualServerField(.imap)
        field.adoptDetected(host: "imap.example.net", security: .startTls)

        #expect(field.port == "143")
        #expect(field.followsSecurity)
    }

    @Test func onlyAnAllDigitTailIsAPort() {
        #expect(ManualServerField.splitHost("imap.example.net").port.isEmpty)
        #expect(ManualServerField.splitHost("imap.example.net:993").host == "imap.example.net")
        #expect(ManualServerField.splitHost("imap.example.net:993").port == "993")
        #expect(ManualServerField.splitHost("host:").port.isEmpty)
        #expect(ManualServerField.splitHost("host:abc").port.isEmpty)
        // Mirrors the core, which splits the same way; a bare IPv6 literal is not a server
        // either side accepts.
        #expect(ManualServerField.splitHost("::1").host == ":")
    }
}
